fdbrpc/include/fdbrpc/AsyncFileWriteChecker.actor.h

```diff
@@ -0,0 +1,381 @@
+			}
+			if (maxFullPage > newMaxFullPage) {
+				maxFullPage = newMaxFullPage;
+			}
+		}
+
+		int size() { return m.size(); }
+
+		bool exist(uint32_t page) { return m.find(page) != m.end(); }
+
+		WriteInfo find(uint32_t page) {
+			if (!exist(page)) {
+				TraceEvent(SevError, "CheckerTryFindingPageNotExist").log();
+				return WriteInfo();
+			}
+			return m[page]->writeInfo;
+		}
+
+		uint32_t leastRecentlyUsedPage() {
+			if (m.size() == 0) {
+				return -1;
+			}
+			return end->prev->page;
+		}
+
+		void remove(uint32_t page) {
+			if (m.find(page) == m.end()) {
+				return;
+			}
+			node* n = m[page];
+			removeFromList(n);
+			m.erase(page);
+			delete n;
+		}
+	};
+
+	AsyncFileWriteChecker(Reference<IAsyncFile> f) : m_f(f), lru("aa") {
+		// Initialize the static history budget the first time (and only the first time) a file is opened.
+		if (!checksumHistoryBudget.present()) {
+			checksumHistoryBudget = FLOW_KNOBS->PAGE_WRITE_CHECKSUM_HISTORY;
+		}
+		totalCheckedSucceed = 0;
+		totalCheckedFail = 0;
+		lru = LRU(m_f->getFilename());
+	}
+
+	~AsyncFileWriteChecker() override { checksumHistoryBudget.get() += lru.size(); }
+
+private:
+	// transform from unixTime(double) to uint32_t, to retain ms precision.
+	static uint32_t transformTime(double unixTime) { return (uint32_t)(unixTime * millisecondsPerSecond); }
+	Reference<IAsyncFile> m_f;
+	Future<Void> checksumWorker;
+	Future<Void> checksumLogger;
+	LRU lru;
+
+	uint64_t totalCheckedFail, totalCheckedSucceed;
+	uint32_t syncedTime;
+	bool checksumWorkerStart = false;
+	std::unordered_set<uint32_t> writing;
+	// This is the most page checksum history blocks we will use across all files.
+	static Optional<int> checksumHistoryBudget;
+	static int checksumHistoryPageSize;
+
+	ACTOR static Future<Void> sweep(AsyncFileWriteChecker* self) {
+		state void* data = (void*)new char[checksumHistoryPageSize];
+		loop {
+			// for each page, read and do checksum
+			// scan from the least recently used, thus it is safe to quit if data has not been synced
+			state uint32_t page = self->lru.leastRecentlyUsedPage();
+			while (self->writing.find(page) != self->writing.end()) {
+				// avoid concurrent ops
+				wait(delay(3.0));
+				continue;
+			}
+			if (page == -1) {
+				// no more available pages
+				break;
+			}
+			int64_t offset = page * checksumHistoryPageSize;
+			// perform a read to verify checksum
+			wait(success(self->read(data, checksumHistoryPageSize, offset)));
+			self->lru.remove(page);
+		}
+		delete[] reinterpret_cast<char*>(data);
+		return Void();
+	}
+
+	ACTOR Future<Void> runChecksumWorker(AsyncFileWriteChecker* self) {
+		// periodically scan the LRU of checksum history, remove them after check the items that have been synced
+		loop {
+			choose {
+				when(wait(AsyncFileWriteChecker::sweep(self))) {}
+				when(wait(delay(FLOW_KNOBS->ASYNC_FILE_WRITE_CHEKCER_CHECKING_INTERVAL))) {}
+			}
+			wait(delay(FLOW_KNOBS->ASYNC_FILE_WRITE_CHEKCER_CHECKING_INTERVAL));
```

Comment:
i see, yeah, that makes sense, I might have mis-understood the requirement, we can remove this sleep, and let the `sweep` keep running, and it will sleep for a shorter time when there is no page available, and then resume.